calcom/cal.com
# feat: add opt-in ready-to-deploy message queue (QStash+Next.js functions)

## What does this PR do?



This PR introduces **a robust message queuing system that is tightly integrated in the codebase and very extensible**. It is a fully opt-in feature for production-like cal.com deployments that want to enhance the scalability, reliability, and performance of their application. It has already been integrated for all emails and webhooks.

![image](https://github.com/calcom/cal.com/assets/6707437/bff397d3-5297-4965-be6c-3e621062090b)


## Benefits:

- Extensibility: Fully opt-in, designed to cater to diverse deployment scenarios.
- Efficiency: Simplifies email dispatch processes with unified parameter treatment, less code bloat.
- Minimal code and infrastructure overhead (less than 500 lines with emails and webhook integrations) and uses mostly existing infrastructure
- Enhanced Security: Robust token-based message sending and signature checks for task runners ensure high security.

## Short Technical Video and Demo

[Short Technical Video and Demo](https://youtu.be/6NhpcW8Rqok?si=0hC_2kLH_IN3EJeQ)


## More info

[Longer Video Description](https://youtu.be/0C8qExHmmg8?si=pMI9mAl-5U9pUCBT)

This PR also showcases the integration for:
- [DONE] All emails send in the application by means of substituting the sendEmail calls with a dispatcher (we unified the parameters in a single object to simplify the treatment)
- [DONE] All Webhook handling
- [FUTURE] Integration with 3rd Party Calendar Event APIs could also be done

Even if done as part of OSShack to address https://osshack.com/bounty/25 (aka https://github.com/calcom/cal.com/issues/12556 or  [CAL-2764](https://linear.app/calcom/issue/CAL-2764/implement-message-queuing-system)), the architecture and  technological choices were purposely done to minimise the amount of changes required and ensure minimal bloating and required changes. This changes will also improve the reliability and the performance of the cal.com app in general.

We only add a single lightweight external dependency ([Upstash QStash](https://upstash.com/docs/qstash/overall/getstarted)) as a message queue from a provider (Upstash) that is already used for another another feature in the code base (rate limiting). Furthermore, any similar queue system that has the capability to call URLs could also be integrated quite easily as an alternative.

As we use application functions as task runners, the infrastructure overhead is minimal and everything is integrated in the same code base. Because in production settings Next.js functions are serveless the system has great scalability. Similarly, existing tools for logging and performance monitoring (i.e. Sentry, Vercel logging and others) will work readily. In addition, there is an additional logging service for the message queue that is easily accesible from the QStash dashboard.

For testing, as the task runners for the different types of tasks are simply routes so they can be tested as any other Next.js pages route (app directory routes could also be used in the future). From a security standpoint, the sending of messages to the queue requires a token while the task runner checks the signature checking both the origin and message tampering, hence it is is quite solid.



## Requirement/Documentation



- See requirements in [CAL-2764](https://linear.app/calcom/issue/CAL-2764/implement-message-queuing-system) and https://github.com/calcom/cal.com/issues/12556 

## Type of change




## How should this be tested?



For end to end local testing an ngrok or similar tunnel is required. But if `QSTASH_URL=localhost` the task runners can be tested locally as any other Next.js function and also by means of using the app. 

- Are there environment variables that should be set?

To test the QStash integration, `QSTASH_URL` and `QSTASH_TOKEN` must be set.

- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Mandatory Tasks


## Checklist



- We have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code does follow the style guidelines of this project
- We have commented my code, particularly in hard-to-understand areas
- We have checked if my PR needs changes to the documentation to a certain degree
- We have checked if my changes generate no new warnings to a certain degree
- We haven't added tests abundant tests yet
- We haven't checked if new and existing unit tests pass locally with my changes
